crates/hotshot-stake-table/src/vec_based.rs

```diff
@@ -0,0 +1,436 @@
+    }
+
+    fn try_iter(&self, version: SnapshotVersion) -> Result<Self::IntoIter, StakeTableError> {
+        match version {
+            SnapshotVersion::Head => Ok(self.head.clone().into_iter()),
+            SnapshotVersion::EpochStart => Ok(self.epoch_start.clone().into_iter()),
+            SnapshotVersion::LastEpochStart => Ok(self.last_epoch_start.clone().into_iter()),
+            SnapshotVersion::BlockNum(_) => Err(StakeTableError::SnapshotUnsupported),
+        }
+    }
+}
+
+impl<K, F> StakeTable<K, F>
+where
+    K: Eq + Hash + Clone + ToFields<F>,
+    F: RescueParameter,
+{
+    /// Initiating an empty stake table.
+    /// Overall capacity is `TREE_BRANCH.pow(height)`.
+    pub fn new() -> Self {
+        let comm = RescueCRHF::sponge_with_zero_padding(&[], 1)[0];
+        Self {
+            head: vec![],
+            epoch_start: vec![],
+            last_epoch_start: vec![],
+            head_total_stake: U256::zero(),
+            epoch_start_total_stake: U256::zero(),
+            last_epoch_start_total_stake: U256::zero(),
+            mapping: HashMap::new(),
+            epoch_start_comm: comm,
+            last_epoch_start_comm: comm,
+        }
+    }
+
+    /// Update the stake table when the epoch number advances, should be manually called.
+    pub fn advance(&mut self) {
+        // Could we avoid this `clone()`?
+        self.last_epoch_start = self.epoch_start.clone();
+        self.last_epoch_start_total_stake = self.epoch_start_total_stake;
+        self.last_epoch_start_comm = self.epoch_start_comm;
+        self.epoch_start = self.head.clone();
+        self.epoch_start_total_stake = self.head_total_stake;
+        self.epoch_start_comm = self.compute_head_comm();
+    }
+
+    /// Set the stake withheld by `key` to be `value`.
+    /// Return the previous stake if succeed.
+    pub fn set_value(&mut self, key: &K, value: U256) -> Result<U256, StakeTableError> {
+        match self.mapping.get(key) {
+            Some(pos) => {
+                let old_value = self.head[*pos].1;
+                self.head[*pos].1 = value;
+                self.head_total_stake -= old_value;
+                self.head_total_stake += value;
+                Ok(old_value)
+            }
+            None => Err(StakeTableError::KeyNotFound),
+        }
+    }
+
+    /// Helper function to recompute the stake table commitment for head version
+    fn compute_head_comm(&mut self) -> F {
+        if self.head.is_empty() {
+            return RescueCRHF::sponge_with_zero_padding(&[], 1)[0];
+        }
+        let mut to_be_hashed = vec![];
+        self.head.iter().for_each(|(key, amount)| {
+            to_be_hashed.extend(key.to_fields());
+            to_be_hashed.push(u256_to_field(amount));
+        });
+        let mut comm = to_be_hashed[0];
+        for i in (1..self.head.len()).step_by(2) {
+            comm = RescueCRHF::sponge_with_zero_padding(
```

Comment:
we shouldn't use this lower-level API, we should use `VariableLengthRescueCRHF::evaluate` instead.
(same for the usage a couple of lines above)